File: Zigzag Conversion.java
// Leetcode Problem 6: Zigzag Conversion
// Java Code
class Solution {
    public String convert(String s, int numRows) {
        if (numRows == 1 || numRows >= s.length()) {
            return s;
        }

        StringBuilder[] rows = new StringBuilder[numRows];
        for (int i = 0; i < numRows; i++) {
            rows[i] = new StringBuilder();
        }

        int currentRow = 0;
        boolean goingDown = true;

        for (char c : s.toCharArray()) {
            rows[currentRow].append(c);

            if (currentRow == 0) {
                goingDown = true;
            } else if (currentRow == numRows - 1) {
                goingDown = false;
            }

            currentRow += goingDown ? 1 : -1;
        }

        StringBuilder result = new StringBuilder();
        for (StringBuilder row : rows) {
            result.append(row);
        }

        return result.toString();
    }
}

File: Zigzag Conversion.py
# Leetcode Problem 6: Zigzag Conversion
# Python Code
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1 or numRows >= len(s):
            return s

        rows = [''] * numRows
        currentRow = 0
        goingDown = True

        for c in s:
            rows[currentRow] += c

            if currentRow == 0:
                goingDown = True
            elif currentRow == numRows - 1:
                goingDown = False

            currentRow += 1 if goingDown else -1

        return ''.join(rows)